### PR TITLE
Handle CsmIndent and CsmUnindent in the interpret method

### LINKED ISSUE #1321: UnsupportedOperationException: CsmIndent
Parser-Version: 3.5.5 (Shipped by Java-Symbol-Solver 0.6.3)

In my application I am processing a piece of legacy Code and would like to modify it. Therefore I parse the Code using the Java-Parser (and Symbol-Solver) and trying to replace the (NodeList) content of an BlockStatement. The Code looks as follows:

```
public void consume(final List<Node> pNodesToProcess, final Optional<FieldDeclaration> p) {
        // Workaround to add multiple children from same node in LexicalPreserving printer => Save former parent (Issue #358 on Java-Symbol-Solver project)
        Optional<BlockStmt> parent = Optional.empty();
        for (final Node n : pNodesToProcess) {
            parent = parent.isPresent() ? parent : n.findParent(BlockStmt.class);
            parent.ifPresent(surroundingBlock -> {
                final NodeList<Statement> newStatements = new NodeList<>();
                Iterator<Node> iterator = surroundingBlock.getChildNodes().iterator();
                while (iterator.hasNext()) {
                    Statement next = (Statement) iterator.next();
                    Statement child = copy(next);
                    newStatements.add(child);
                    if (pNodesToProcess.contains(child)) {
                        process(p, newStatements, child);
                    }
                }
                surroundingBlock.replace(new BlockStmt(newStatements)); **// Errornous line #1**
            });
        }
    }

    private void process(final Optional<FieldDeclaration> p, final NodeList<Statement> newStatements, Statement child) {
        final Optional<MethodCallExpr> firstContainedMethodCall = child.findFirst(MethodCallExpr.class);
        if (firstContainedMethodCall.isPresent()) {
            final MethodCallExpr methodCallExpr = firstContainedMethodCall.get().clone();
            Expression expression = methodCallExpr.getScope().get();
            **// Potential beginning of error causing Code #2**
            if (expression.isNameExpr()) {
                expression.asNameExpr().setName(p.get().findFirst(VariableDeclarator.class).get().getNameAsString());
            } else if (expression.isFieldAccessExpr()) {
                expression.asFieldAccessExpr().getName().setIdentifier(p.get().findFirst(VariableDeclarator.class).get().getNameAsString());
            }
            newStatements.add(new ExpressionStmt(methodCallExpr));
            **// end of error causing code**
        }
    }

    private Statement copy(final Statement next) {
        if (next.isExpressionStmt()) {
            ExpressionStmt expressionStmt = new ExpressionStmt(next.asExpressionStmt().getExpression());
            next.getRange().ifPresent(r -> {
                expressionStmt.setRange(r);
            });
            return expressionStmt;
        }
        return next.clone();
    }
```

When executing the Code I get following stacktrace:

```
java.lang.UnsupportedOperationException: CsmIndent
	at com.github.javaparser.printer.lexicalpreservation.LexicalPreservingPrinter.interpret(LexicalPreservingPrinter.java:392)
	at com.github.javaparser.printer.lexicalpreservation.LexicalPreservingPrinter.prettyPrintingTextNode(LexicalPreservingPrinter.java:366)
	at com.github.javaparser.printer.lexicalpreservation.LexicalPreservingPrinter.getOrCreateNodeText(LexicalPreservingPrinter.java:417)
	at com.github.javaparser.printer.lexicalpreservation.ChildTextElement.getNodeTextForWrappedNode(ChildTextElement.java:56)
	at com.github.javaparser.printer.lexicalpreservation.TextElementIteratorsFactory.reverseIterator(TextElementIteratorsFactory.java:170)
	at com.github.javaparser.printer.lexicalpreservation.TextElementIteratorsFactory.partialReverseIterator(TextElementIteratorsFactory.java:184)
	at com.github.javaparser.printer.lexicalpreservation.TextElementIteratorsFactory.reverseIterator(TextElementIteratorsFactory.java:178)
	at com.github.javaparser.printer.lexicalpreservation.TextElementIteratorsFactory.reverseIterator(TextElementIteratorsFactory.java:171)
	at com.github.javaparser.printer.lexicalpreservation.TextElementIteratorsFactory.partialReverseIterator(TextElementIteratorsFactory.java:184)
	at com.github.javaparser.printer.lexicalpreservation.LexicalPreservingPrinter.tokensPreceeding(LexicalPreservingPrinter.java:292)
	at com.github.javaparser.printer.lexicalpreservation.LexicalPreservingPrinter.findIndentation(LexicalPreservingPrinter.java:425)
	at com.github.javaparser.printer.lexicalpreservation.Difference.apply(Difference.java:488)
	at com.github.javaparser.printer.lexicalpreservation.LexicalDifferenceCalculator.calculatePropertyChange(LexicalDifferenceCalculator.java:121)
	at com.github.javaparser.printer.lexicalpreservation.LexicalPreservingPrinter$2.concretePropertyChange(LexicalPreservingPrinter.java:185)
	at com.github.javaparser.ast.observer.PropagatingAstObserver.propertyChange(PropagatingAstObserver.java:62)
	at com.github.javaparser.ast.Node.lambda$notifyPropertyChange$3(Node.java:537)
	at java.util.ArrayList.forEach(ArrayList.java:1249)
	at com.github.javaparser.ast.Node.notifyPropertyChange(Node.java:537)
	at com.github.javaparser.ast.body.MethodDeclaration.setBody(MethodDeclaration.java:137)
	at com.github.javaparser.ast.body.MethodDeclaration.replace(MethodDeclaration.java:309)
	at com.github.javaparser.ast.Node.replace(Node.java:506)
	at de.refactoring.ExtendingAction.lambda$0(ExtendingAction.java:38)
	at java.util.Optional.ifPresent(Optional.java:159)
	at de.refactoring.ExtendingAction.consume(ExtendingAction.java:27)
	at de.refactoring.MethodCallChange.lambda$0(MethodCallChange.java:57)
	at java.lang.Iterable.forEach(Iterable.java:75)
	at de.refactoring.MethodCallChange.apply(MethodCallChange.java:52)
	at de.refactoring.MethodCallChangeTest.extendingActionGivesExpectedResult(MethodCallChangeTest.java:175)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
	at org.junit.vintage.engine.execution.RunnerExecutor.execute(RunnerExecutor.java:42)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:184)
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
	at java.util.Iterator.forEachRemaining(Iterator.java:116)
	at java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
	at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:418)
	at org.junit.vintage.engine.VintageTestEngine.executeAllChildren(VintageTestEngine.java:83)
	at org.junit.vintage.engine.VintageTestEngine.execute(VintageTestEngine.java:74)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:170)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:154)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:90)
	at org.junit.platform.runner.JUnitPlatform.run(JUnitPlatform.java:133)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:86)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:38)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:459)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:678)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:382)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:192)
```
I set up my CompilationUnit by following Code:
```
    // Create type solver from CodeBase
    final TypeSolver mySolver = TypeSolverFactory.createFrom(codeBase);

    // Create new preserving compilation unit
    Pair<File, CompilationUnit> cu = CompilationUnitFactory.createPreservingCompilationUnit(next, mySolver);
```

which in fact calls **`LexicalPreservingPrinter.setup(JavaParser.parse(pFromFile))`** in the background.

What is weird I can execute the line tagged as #1 when I comment out Code snippet #2 (other way round as well of course). Hope you guys can have a look on it :)
Maybe the issue could also be somehow related to issue #358 from Java-Symbol-Solver project.

### PR DESCRIPTION
**Pull request by ftomassetti**
Fix #1321 handling CsmIndent and CsmUnindent in the interpret method

### CODE REVIEW

**Comment by matozoid** [comment]
Oh, this one is very welcome!

File: javaparser-core-testing/src/test/java/com/github/javaparser/printer/lexicalpreservation/LexicalPreservingPrinterTest.java [hunk 1]
```diff
@@ -1104,4 +1106,16 @@ public void addedLineCommentsPrinted() {
...
+
+    @Test
+    public void issue1321() {
+        CompilationUnit compilationUnit = JavaParser.parse("class X { X() {} private void testme() {} }");
+        LexicalPreservingPrinter.setup(compilationUnit);
+
+        ClassOrInterfaceDeclaration type = compilationUnit.getClassByName("X").get();
+        type.getConstructors().get(0).setBody(new BlockStmt().addStatement("testme();"));
+
+        assertEquals("class X { X() {\n    testme();\n} private void testme() {} }", LexicalPreservingPrinter.print(compilationUnit));
```
Comment (matozoid) [inline javaparser-core-testing/src/test/java/com/github/javaparser/printer/lexicalpreservation/LexicalPreservingPrinterTest.java]: AppVeyor does its job: this test fails on Windows. Try `assertEqualsNoEol` or whatever it is called.

**Comment by ftomassetti** [comment]
The most important part was made by @matf who took the time to write a simple test

File: javaparser-core/src/main/java/com/github/javaparser/printer/lexicalpreservation/LexicalPreservingPrinter.java [hunk 2]
```diff
@@ -420,6 +418,9 @@ private static void prettyPrintingTextNode(Node node, NodeText nodeText) {
         interpret(node, ConcreteSyntaxModel.forClass(node.getClass()), nodeText);
     }
 
+    /**
+     * TODO: Process CsmIndent and CsmUnindent before reaching this point
+     */
```
Comment (matozoid) [inline javaparser-core/src/main/java/com/github/javaparser/printer/lexicalpreservation/LexicalPreservingPrinter.java:423]: I guess we need a new issue for this then.

File: javaparser-core-testing/src/test/java/com/github/javaparser/printer/lexicalpreservation/LexicalPreservingPrinterTest.java (hunk 1 above)
Comment (ftomassetti) [reply javaparser-core-testing/src/test/java/com/github/javaparser/printer/lexicalpreservation/LexicalPreservingPrinterTest.java]: Right, it slipped in, sorry about that

**Comment by matf** [comment]
Cool, thanks for taking the time to look into this!

**Comment by matozoid** [comment]
Thanks @matf!

**Pull request merged**